analyzers/copyboundary/copy_boundary.go

```diff
@@ -0,0 +1,100 @@
+// Package copyboundary defines an Analyzer that checks for
+// uses of copy of slice & map.
+// https://github.com/uber-go/guide/blob/master/style.md#copy-slices-and-maps-at-boundaries
+package copyboundary
+
+import (
+	"go/ast"
+
+	"golang.org/x/tools/go/analysis"
+)
+
+const Doc = `check for uses of copy of slice & map
+
+This checker reports usages of slice & map that are directly copied. 
+Slices and maps contain pointers to the underlying data so be wary 
+of scenarios when they need to be copied..
+`
+
+var Analyzer = &analysis.Analyzer{
+	Name: "copyboundary",
+	Doc:  Doc,
+	Run:  run,
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	inspector := newInspector(pass)
+	for _, file := range pass.Files {
+		for _, inspect := range inspector.collections {
+			ast.Inspect(file, inspect)
+		}
+	}
+	return nil, nil
+}
+
+type copyBoundaryInspector struct {
+	pass        *analysis.Pass
+	collections []func(node ast.Node) bool
+}
+
+func newInspector(pass *analysis.Pass) *copyBoundaryInspector {
+	inspectors := &copyBoundaryInspector{
+		pass: pass,
+	}
+	inspectors.collections = []func(ast.Node) bool{
+		inspectors.inspectAssignStatement,
+	}
+	return inspectors
+}
+
+func (i *copyBoundaryInspector) inspectAssignStatement(node ast.Node) bool {
+	assignStmt, ok := node.(*ast.AssignStmt)
+	if !ok {
+		return true
+	}
+
+	typ := checkType(assignStmt.Lhs)
+	if typ == "" {
+		return true
+	}
+
+	typ = checkType(assignStmt.Rhs)
+	if typ == "" {
+		return true
+	}
+
+	i.pass.Reportf(assignStmt.Pos(), "copy-boundary: copies a %s directly", typ)
+
+	return false
+}
+
+func checkType(obj []ast.Expr) string {
+	if len(obj) > 1 {
+		return ""
+	}
+
+	x, ok := obj[0].(*ast.Ident)
+	if !ok {
+		return ""
+	}
+
+	var typ ast.Expr
+
+	switch decl := x.Obj.Decl.(type) {
```

Comment:
This will panic. @sadlil

Reply:
I've tested the code using this:
```
data := []int{1,2,3}
dataCopied := data
```
It didn't panic.

Reply:
share your sample code. just this part

Reply:
```
func testDeclareSliceFunc() {
	data := []int{1, 2, 3}
	dataCopied := data
	fmt.Print(dataCopied)
}
```

Actually I added it to the test code. And ran the test. The test passed successfully.

Reply:
Make a binary `gouberfmt`. Tun that to check this snippet directly.

Reply:
Tried it. Didn't panic.

```golang
    x, ok := obj[0].(*ast.Ident)
	if !ok {
		return ""
	}
```

As the function is returned if it's not a `*ast.Ident` type, It shouldn't panic, right?

Reply:
`_ = dataCopied` 
This causes panic

Reply:
oops. This makes the `Obj` nil, my bad. I will fix it in the next PR. Thanks for reviewing it!